### PR TITLE
tests: add tests for UDP functions

### PR DESCRIPTION
**Pull request by authmillenon**
Depends on #3248 and ~~#3295~~ (merged).

### CODE REVIEW

File: sys/net/transport_layer/ng_udp/ng_udp.c [hunk 1]
```diff
@@ -39,7 +39,7 @@
 /**
  * @brief   Save the UDP's thread PID for later reference
  */
-static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+kernel_pid_t ng_udp_pid = KERNEL_PID_UNDEF;
```
Comment (haukepetersen) [inline sys/net/transport_layer/ng_udp/ng_udp.c]: I don't like this change... Feels wrong to expose more internals just for testing. Exposing this leads potentially to people miss-using this... Why don't you just use `netreg` in the test's for lookup?

Comment (kaspar030) [inline sys/net/transport_layer/ng_udp/ng_udp.c]: > I don't like this change...

Same feeling here.

Comment (authmillenon) [inline sys/net/transport_layer/ng_udp/ng_udp.c]: Can we continue this discussion in #3248 which introduced this change?

**Comment by OlegHahm** [comment]
I think, what's urgently needed before we write any automatically executable tests outside the unittests folder is an extension of the CI system to actually run these tests. Otherwise, having the certainty that some of the stuff is *not* tested, seems better to me than having the false illusion of tested code where the tests in reality become never executed.

Several times in the past, I was relying on the fact that the CI system will automatically check and test stuff, that was actually not the case. (Such as build warnings.)

**Comment by authmillenon** [comment]
The unittests are already quite heavy loaded and do not run on the majority of the boards (though you can run them seperately). Adding comparably huge modules with whole threads doesn't soften this fact. For the `netapi` modules it is especially precarious to test them unisolated, since compiling them with other modules makes there functionality more complex, making them harder to test, making it absolutely senseless to put them in different modules in the first place. Sure it is not certain iff the stuff is tested, but careful review *can* assure that.

**Comment by OlegHahm** [comment]
> The unittests are already quite heavy loaded and do not run on the majority of the boards (though you can run them seperately). 

I agree and I don't say this should be integrated into unittests. I just want to make sure, that we actually ensure that we run our tests automatically more or less periodically.

> For the netapi modules it is especially precarious to test them unisolated, since compiling them with other modules makes there functionality more complex, making them harder to test, making it absolutely senseless to put them in different modules in the first place.

I think in the long we run, we need to mock some stuff, but for now, you're right.

**Comment by authmillenon** [comment]
> I think in the long we run, we need to mock some stuff, but for now, you're right.

For netapi modules there is [`ng_nettest`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ng_nettest.h) already ;-)

**Comment by PeterKietzmann** [comment]
@A-Paul may I assign you?

**Comment by A-Paul** [comment]
@PeterKietzmann: Did it already.

**Comment by A-Paul** [comment]
@authmillenon, what's your opinion on following suggestions:

+ If I didn't miss something, #3248 is full contained in this PR. Can #3248 be canceled and we just continue working on this?
+ I think it's sensible to rebase and refactor, before I place questions.
+ I already like the function comments you did in `udp-netapi-tests.c` ( the existence of `\pre` and `\post` tag was new to me) can you do that for `udp-func-tests.c` too?

In the end this is a valuable addition and will be merged somehow, so you won't do it to no purpose.

**Comment by authmillenon** [comment]
There are two ways to handle depending PRs inferred from there current state:

1. There is no work need to be done on upstream PR: Just ACK the downstream PR and merge it, the upstream PR will be marked as merged as soon as its commits are in master too.
2. The upstream PR needs reviewing too: Review the upstream PR first and merge it separately (it is for the most part easier, because it contains less changes). Then review the downstream PR.

As long as both PRs aren't very simple (which those two here are not) I always would choose 2. and ignore "Waiting for other PR" labeled PRs in a similar fassion as WIP PRs.

**Comment by A-Paul** [comment]
@authmillenon, accepted. If you prefer to maintain to PRs.

> + I think it's sensible to rebase and refactor, before I place questions.

This is obsolete. I was lazy and just looked for a "Rebase comment" which isn't there. But I've seen you rebased already.

**Comment by authmillenon** [comment]
Closed for same reason as in https://github.com/RIOT-OS/RIOT/pull/3248#issuecomment-174184708

**Comment by kYc0o** [comment]
any news?

**Comment by authmillenon** [comment]
Yes, it's closed, due to the planned test system overhaul ;)

**Comment by kYc0o** [comment]
Ok, I didn't see that it was closed, thanks ;)

**Pull request closed without merging**